src/logic/voice/voice.ts

```diff
@@ -0,0 +1,132 @@
+import { DisconnectReason } from '@livekit/protocol'
+import { AppComponents } from '../../types'
+import { IVoiceComponent } from './types'
+import { LivekitCredentials } from '../../types/livekit.type'
+import { getPrivateVoiceChatRoomName } from './utils'
+
+export function createVoiceComponent(components: Pick<AppComponents, 'voiceDB' | 'logs' | 'livekit'>): IVoiceComponent {
+  const { voiceDB, livekit, logs } = components
+  const logger = logs.getLogger('voice')
+
+  /**
+   * Handles the event when a participant joins a room.
+   * @param userAddress - The address of the user that joined the room.
+   * @param roomName - The name of the room the user joined.
+   */
+  async function handleParticipantJoined(userAddress: string, roomName: string) {
+    // Check if the room is active.
+    const isRoomActive = await voiceDB.isPrivateRoomActive(roomName)
+    if (!isRoomActive) {
+      logger.warn(`User ${userAddress} has joined an inactive room ${roomName}, destroying it`)
+
+      // Destroy the room to disconnect the user.
+      return livekit.deleteRoom(roomName)
+    }
+
+    // Join the user to the room.
+    const { oldRoom } = await voiceDB.joinUserToRoom(userAddress, roomName)
+
+    // If the user was in another room, destroy the old room.
+    if (oldRoom !== roomName) {
+      await livekit.deleteRoom(oldRoom)
+    }
+  }
+
+  /**
+   * Handles the event when a participant leaves a room.
+   * @param userAddress - The address of the user that left the room.
+   * @param roomName - The name of the room the user left.
+   * @param disconnectReason - The reason the user left the room.
+   */
+  async function handleParticipantLeft(userAddress: string, roomName: string, disconnectReason: DisconnectReason) {
+    // If the user disconnected because of a duplicate identity, do nothing. They're re-joining the room.
+    if (disconnectReason === DisconnectReason.DUPLICATE_IDENTITY) {
+      return
+    }
+
+    if (disconnectReason === DisconnectReason.CLIENT_INITIATED) {
```

Comment:
Curious about what does `CLIENT_INITIATED` means as a disconnect reason?

Reply:
Sure! The `CLIENT_INITIATED` status means that the client disconnected willingly from the room, that is, they have issued an explicit disconnection. I'll improve the comments so it's clearer.